runtime/tests/immediateString.c

```diff
@@ -0,0 +1,53 @@
+#include "stringUtils.h"
+
+HASH_DEFINE_KEY;
+
+void validateImmediateStringLength(uint64_t bits, int nChars, int nCodePoints) {
+	StringLength length = immediateStringLength(bits);
+	assert(length.nBytes == nChars);
+	assert(length.nCodePoints == nCodePoints);
+	StringLength taggedLength = taggedStringLength(bitsToTaggedPtr(bits));
+	assert(taggedLength.nBytes == length.nBytes);
+	assert(taggedLength.nCodePoints == length.nCodePoints);
+}
+
+void validateTaggedStringEquals(uint64_t lhs, uint64_t rhs, bool expected) {
+	TaggedPtr lhsPtr = bitsToTaggedPtr(lhs);
+	TaggedPtr rhsPtr = bitsToTaggedPtr(rhs);
+	bool result = taggedStringEqual(lhsPtr, rhsPtr);
+	assert(result == expected);
+}
+
+void testImmediateStringLength() {
+	// test ASCII
+	validateImmediateStringLength(0xABFFFFFFFFFFFFFF, 0, 0); // ""
+	validateImmediateStringLength(0xABFFFFFFFFFF4142, 2, 2); // "AB"
+	validateImmediateStringLength(0xABFFFFFFFF613223, 3, 3); // "a2#"
+	validateImmediateStringLength(0xABFFFFFF256C6420, 4, 4); // "%ld "
+	validateImmediateStringLength(0xABFFFF6162606364, 5, 5); // "ab`cd"
+	validateImmediateStringLength(0xABFF2A2625242C3F, 6, 6); // "*&%$,?"
+	validateImmediateStringLength(0xAB61626364656667, 7, 7); // "abcdefg"
+	// testUnicode 
```

Comment:
@jclark I think there is a bug in `immediateStringLength` with unicode characters. I have commented the wrong values I got here.

Reply:
This is not correct UTF-8. In the UTF-8 encoding of a non-ASCII character, all bytes have the top bit set and the first byte has the top 2 bits set. https://en.wikipedia.org/wiki/UTF-8#Encoding

In 0xABFFFFFFFFFF265E, 0x5E is supposed to be the first byte but it does not have the top bit set.

And why is the top byte 0xAB. It should be 0x2A, shouldn't it.

Reply:
I fixed the above problems but I still get the number of bytes wrong (1).

Reply:
The lo byte is still wrong. For a UTF-8 sequence of three bytes, the first byte must be 0b1110xxxx whereas 0x9E is 0b10011110.

Reply:
To be clear if get a utf-8 encoding as `E2` `99` `9E` (for "♞") then it will get encoded as `0x2AFFFFFFFF9E99E2` (first byte as the right most one) not not `0x2AFFFFFFFFE2999E` (first byte as the left most after padding). It work fine in the former case.

Reply:
Yes, the UTF-8 bytes are in little-endian order in the `int`.